docs-website/topics/serialization-customization-options.md

```diff
@@ -0,0 +1,516 @@
+    var id by ::name
+}
+
+fun main() {
+    val data = Project("kotlinx.serialization").apply { stars = 9000 }
+    // Prints only the name and the stars properties
+    println(Json.encodeToString(data))
+    // {"name":"kotlinx.serialization","stars":9000}
+}
+//sampleEnd
+```
+{kotlin-runnable="true" id="serialize-annotation-example"}
+
+### Serialization of class references
+
+Serializable classes can contain properties that reference other classes.
+The referenced classes must also be annotated with `@Serializable`.
+When encoded to JSON, this results in a nested JSON object:
+
+```kotlin
+// Imports declarations from the serialization and JSON handling libraries
+import kotlinx.serialization.*
+import kotlinx.serialization.json.*
+
+//sampleStart
+@Serializable
+// The owner property references another serializable class User
+class Project(val name: String, val owner: User)
+
+// The referenced class must also be annotated with @Serializable
+@Serializable
+class User(val name: String)
+
+fun main() {
+    val owner = User("kotlin")
+    val data = Project("kotlinx.serialization", owner)
+    println(Json.encodeToString(data))
+    // {"name":"kotlinx.serialization","owner":{"name":"kotlin"}}
+}
+//sampleEnd
+```
+{kotlin-runnable="true" id="serialize-class-references"}
+
+> To reference non-serializable classes, annotate the corresponding properties with [`@Transient`](#exclude-properties-with-the-transient-annotation), or
+> provide a [custom serializer](create-custom-serializers.md) for them.
+>
+{style="tip"}
+
+### Serialization of repeated object references
+
+Kotlin Serialization is designed to encode and decode plain data. It doesn't support reconstruction
+of arbitrary object graphs with repeated object references.
+For example, when serializing an object that references the same instance twice, it's encoded twice:
+
+```kotlin
+// Imports declarations from the serialization and JSON handling libraries
+import kotlinx.serialization.*
+import kotlinx.serialization.json.*
+
+//sampleStart
+@Serializable
+class Project(val name: String, val owner: User, val maintainer: User)
+
+@Serializable
+class User(val name: String)
+
+fun main() {
+    val owner = User("kotlin")
+    // References owner twice
+    val data = Project("kotlinx.serialization", owner, owner)
+    println(Json.encodeToString(data))
+    // {"name":"kotlinx.serialization","owner":{"name":"kotlin"},"maintainer":{"name":"kotlin"}}
+}
+//sampleEnd
+```
+{kotlin-runnable="true" id="serialize-repeated-references"}
+
+> If you attempt to serialize a circular structure, it results in stack overflow.
+> To exclude references from serialization, use [the @Transient annotation](#exclude-properties-with-the-transient-annotation).
+>
+{style="tip"}
+
+### Generic class serialization
+
+Generic classes in Kotlin support type-polymorphism, which is enforced by Kotlin Serialization at
+compile-time. For example, consider a generic serializable class `Box<T>`:
+
+```kotlin
```

Comment:
What do you think about this as an alternative? It is the same example with just slightly different names:
```
import kotlinx.serialization.*
import kotlinx.serialization.json.*

// A generic wrapper that can hold various types of data
@Serializable
class Payload<T>(val value: T)

@Serializable
data class Repository(val name: String, val language: String)

@Serializable
class BackupData(
    val issueCount: Payload<Int>,
    val mainRepo: Payload<Repository>
)

fun main() {
    val backup = BackupData(
        issueCount = Payload(42),
        mainRepo = Payload(Repository("kotlinx.serialization", "Kotlin"))
    )
    println(Json.encodeToString(backup))
    // {"issueCount":{"value":42},"mainRepo":{"value":{"name":"kotlinx.serialization","language":"Kotlin"}}}
}
```